### PR TITLE
homework-9

### PR DESCRIPTION
**Pull request by Bloodstinger**
Implementation of linked list

### CODE REVIEW

File: src/lecture/eight/linked/list/CustomLinkedList.java [hunk 1]
```diff
@@ -0,0 +1,193 @@
+package lecture.eight.linked.list;
+
+import lecture.seven.array.list.NegativeIndexException;
+
+public class CustomLinkedList<T> implements List<T> {
+    transient private int listSize;
```
Comment (Sofasmile) [inline src/lecture/eight/linked/list/CustomLinkedList.java]: why do we use transient here?

File: src/lecture/eight/linked/list/CustomLinkedList.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    public CustomLinkedList() {
+    }
+
+    public CustomLinkedList(List<T> list) {
+        this();
+        addAll(list);
+    }
+
+    private static class Node<T> {
+        T item;
```
Comment (Sofasmile) [inline src/lecture/eight/linked/list/CustomLinkedList.java]: add access modifier

File: src/lecture/eight/linked/list/CustomLinkedList.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
+package lecture.eight.linked.list;
+
+import lecture.seven.array.list.NegativeIndexException;
+
+public class CustomLinkedList<T> implements List<T> {
```
Comment (Sofasmile) [inline src/lecture/eight/linked/list/CustomLinkedList.java:8]: please remember about logic organization of the class:  public methods, private methods

File: src/lecture/eight/linked/list/CustomLinkedList.java [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        Object[] result = new Object[listSize];
+        int i = 0;
+        for (Node<T> node = first; node != null; node = node.next) {
+            result[i++] = node.item;
+        }
+        return result;
+    }
+
+    @Override
+    public void add(T value) {
```
Comment (Sofasmile) [inline src/lecture/eight/linked/list/CustomLinkedList.java:60]: better move logic from method linkLast() here

File: src/lecture/eight/linked/list/CustomLinkedList.java [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        last = newNode;
+        if (lastNode == null) {
+            first = newNode;
+        } else {
+            lastNode.next = newNode;
+        }
+        listSize++;
+    }
+
+    private void linkBefore(T item, Node<T> nodeAfter) {
```
Comment (Sofasmile) [inline src/lecture/eight/linked/list/CustomLinkedList.java]: try to use more understandable names

File: src/lecture/eight/linked/list/CustomLinkedList.java [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+    private void checkIndex(int index) {
+        if (index > listSize) {
+            throw new IndexOutOfBoundsException();
+        } else if (index < 0) {
+            throw new NegativeIndexException();
+        }
+    }
+
+    private void linkLast(T item) {
+        final Node<T> lastNode = last;
```
Comment (Sofasmile) [inline src/lecture/eight/linked/list/CustomLinkedList.java]: do we need final here?

File: src/lecture/eight/linked/list/CustomLinkedList.java [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+        Node<T> next;
+
+        Node(Node<T> prev, T object, Node<T> next) {
+            this.item = object;
+            this.prev = prev;
+            this.next = next;
+        }
+    }
+
+    private Node<T> node(int index) {
```
Comment (Sofasmile) [inline src/lecture/eight/linked/list/CustomLinkedList.java]: rename to getNode()

File: src/lecture/eight/linked/list/CustomLinkedList.java [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+            linkLast(value);
+        } else {
+            linkBefore(value, node(index));
+        }
+    }
+
+
+    @Override
+    public void addAll(List<T> list) {
+        Object[] listArray = list.toArray();
```
Comment (Sofasmile) [inline src/lecture/eight/linked/list/CustomLinkedList.java]: try to use method add(T value) to simplify code

File: src/lecture/eight/linked/list/CustomLinkedList.java [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+    public T remove(int index) {
+        checkIndex(index);
+        return unlinkNode(node(index));
+    }
+
+    @Override
+    public T remove(T t) {
+        if (t == null) {
+            for (Node<T> tNode = first; tNode != null; tNode = tNode.next) {
+                if (tNode.item == null) {
```
Comment (Sofasmile) [inline src/lecture/eight/linked/list/CustomLinkedList.java]: maybe better write:
if(Objects.IsNull(tNode.item == null) || t.equals(tNode.item))...

File: src/lecture/eight/linked/list/List.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+    void add(T value, int index);
+
+    void addAll(List<T> list);
+
+    T get(int index);
+
+    void set(T value, int index);
+
+    T remove(int index); // возвращаем элемент, который удалили
```
Comment (Sofasmile) [inline src/lecture/eight/linked/list/List.java]: remove comments please

**Comment by Bloodstinger** [comment]
Added changes due to code review.

File: src/lecture/eight/linked/list/CustomLinkedList.java [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+        } else {
+            linkBeforeNode(value, getNode(index));
+        }
+    }
+
+
+    @Override
+    public void addAll(List<T> list) {
+        Object[] array = list.toArray();
+        for (Object node :
```
Comment (SevostyanovGerman) [inline src/lecture/eight/linked/list/CustomLinkedList.java]: is it ok convension?

**Review by SevostyanovGerman — CHANGES_REQUESTED** [review]

File: src/lecture/eight/linked/list/CustomLinkedList.java [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+    public T remove(T t) {
+        if (t == null) {
+            for (Node<T> tNode = first; tNode != null; tNode = tNode.next) {
+                if (Objects.isNull(tNode.item == null) || t.equals(tNode.item)) {
+                    unlinkNode(tNode);
+                    return t;
+                }
+            }
+        }
+        return null;
```
Comment (SevostyanovGerman) [inline src/lecture/eight/linked/list/CustomLinkedList.java]: return null bad practice

**Review by SevostyanovGerman — CHANGES_REQUESTED** [review]

File: src/lecture/eight/linked/list/CustomLinkedList.java [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+    @Override
+    public T remove(int index) {
+        checkIndex(index);
+        return unlinkNode(getNode(index));
+    }
+
+    @Override
+    public T remove(T t) {
+        if (t == null) {
+            for (Node<T> tNode = first; tNode != null; tNode = tNode.next) {
```
Comment (SevostyanovGerman) [inline src/lecture/eight/linked/list/CustomLinkedList.java]: cs

**Review by SevostyanovGerman — APPROVED** [review]

**Pull request merged**